### PR TITLE
fix(tracing): ensure http.url obfuscation is consistent with other libraries

### PR DESCRIPTION
**Pull request by mabdinur**
When ``DD_TRACE_OBFUSCATION_QUERY_STRING_REGEXP`` is set to an empty string ``global_query_string_obfuscation_disabled`` should be set to False.

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/contrib/trace_utils.py [hunk 1]
```diff
@@ -575,6 +574,27 @@ def activate_distributed_headers(tracer, int_config=None, request_headers=None,
...
+
+    """
+    for lower, upper in error_ranges.split(","):
+        if int(lower) <= status_code <= int(upper):
+            return True
+    return False
+
+
+
+
```
Comment (mabdinur) [inline ddtrace/contrib/trace_utils.py:597]:
```suggestion
```

File: ddtrace/contrib/trace_utils.py [hunk 2]
```diff
@@ -44,6 +44,9 @@
     from ddtrace.settings import IntegrationConfig  # noqa:F401
 
 
+_EMPTY_STRING_PATTERN = re.compile(r"")
+
+
```
Comment (mabdinur) [inline ddtrace/contrib/trace_utils.py:49]:
```suggestion
```

File: ddtrace/contrib/trace_utils.py [hunk 3]
```diff
@@ -404,14 +404,18 @@ def ext_service(pin, int_config, default=None):
...
-    if integration_config.http_tag_query_string:  # Tagging query string in http.url
-        if config.global_query_string_obfuscation_disabled:  # No redacting of query strings
-            span.set_tag_str(http.URL, url)
-        else:  # Redact query strings
-            span.set_tag_str(http.URL, redact_url(url, config._obfuscation_query_string_pattern, query))
-    else:  # Not tagging query string in http.url
+    if not integration_config.http_tag_query_string:  # Tagging query string in http.url
+        span.set_tag_str(http.URL, strip_query_string(url))
+    elif config.global_query_string_obfuscation_disabled:
+        # TODO(munir): deprecate and remove ddtrace.config.global_query_string_obfuscation_disabled,
```
Comment (mabdinur) [inline ddtrace/contrib/trace_utils.py:410]:
```suggestion
        # TODO(munir): Remove ddtrace.config.global_query_string_obfuscation_disabled,
```

File: ddtrace/contrib/trace_utils.py [hunk 4]
```diff
@@ -404,14 +404,18 @@ def ext_service(pin, int_config, default=None):
...
 def _set_url_tag(integration_config, span, url, query):
     # type: (IntegrationConfig, Span, str, str) -> None
-
-    if integration_config.http_tag_query_string:  # Tagging query string in http.url
-        if config.global_query_string_obfuscation_disabled:  # No redacting of query strings
-            span.set_tag_str(http.URL, url)
-        else:  # Redact query strings
-            span.set_tag_str(http.URL, redact_url(url, config._obfuscation_query_string_pattern, query))
-    else:  # Not tagging query string in http.url
+    if not integration_config.http_tag_query_string:  # Tagging query string in http.url
```
Comment (mabdinur) [inline ddtrace/contrib/trace_utils.py:407]:
```suggestion
    if not integration_config.http_tag_query_string:
```

File: tests/tracer/test_env_vars.py [hunk 5]
```diff
@@ -8,7 +8,7 @@
     "env_var_name,env_var_value,expected_obfuscation_config,expected_global_query_string_obfuscation_disabled,"
     "expected_http_tag_query_string",
     [
-        ("DD_TRACE_OBFUSCATION_QUERY_STRING_REGEXP", "", None, True, True),
+        ("DD_TRACE_OBFUSCATION_QUERY_STRING_REGEXP", "", 're.compile("".encode("ascii"))', True, True),
```
Comment (mabdinur) [inline tests/tracer/test_env_vars.py:11]:
```suggestion
        ("DD_TRACE_OBFUSCATION_QUERY_STRING_REGEXP", "", 're.compile(b"")', True, True),
```

File: tests/tracer/test_trace_utils.py [hunk 6]
```diff
@@ -1001,6 +1001,57 @@ def test_sanitized_url_in_http_meta(span, int_config):
...
+            span,
+            config.myint,
+            method="GET",
+            url=SENSITIVE_URL,
+            status_code=200,
+        )
+        assert span.get_tag(http.URL) == SENSITIVE_URL
+
+
+# TODO(munir): Remove this test when global_query_string_obfuscation_disabled is deprecated/removed
```
Comment (mabdinur) [inline tests/tracer/test_trace_utils.py:1026]:
```suggestion
# TODO(munir): Remove this test when global_query_string_obfuscation_disabled is removed
```

File: releasenotes/notes/make-obfuscation-consistent-8a2361bf3298e1be.yaml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+---
+fixes:
+  - |
+    tracing: Ensures that ``http.url`` span tags contain the full query string when ``DD_TRACE_OBFUSCATION_REGEX`` is set to an empty string.
```
Comment (mabdinur) [inline releasenotes/notes/make-obfuscation-consistent-8a2361bf3298e1be.yaml:4]:
```suggestion
    tracing: Ensures that ``http.url`` span tag contains the full query string when ``DD_TRACE_OBFUSCATION_QUERY_STRING_REGEXP `` is set to an empty string.
```

File: ddtrace/contrib/trace_utils.py [hunk 8]
```diff
@@ -404,14 +404,18 @@ def ext_service(pin, int_config, default=None):
...
-    else:  # Not tagging query string in http.url
+    if not integration_config.http_tag_query_string:
+        span.set_tag_str(http.URL, strip_query_string(url))
+    elif config.global_query_string_obfuscation_disabled:
+        # TODO(munir): Remove this case when config.global_query_string_obfuscation_disabled is removed (v3.0),
+        # setting config._obfuscation_query_string_pattern to an empty string should be enough to disable obfuscation
+        span.set_tag_str(http.URL, url)
+    elif getattr(config._obfuscation_query_string_pattern, "pattern", None) == b"":
+        # if obfuscation is enabled but the pattern is empty, this means config.global_query_string_obfuscation_disabled
+        # was manually set to False, so we should strip the whole query string
```
Comment (mabdinur) [inline ddtrace/contrib/trace_utils.py:415]:
```suggestion
        # obfuscation is disabled when DD_TRACE_OBFUSCATION_QUERY_STRING_REGEXP="" 
```

File: ddtrace/contrib/trace_utils.py [hunk 9]
```diff
@@ -404,14 +404,18 @@ def ext_service(pin, int_config, default=None):
...
-        if config.global_query_string_obfuscation_disabled:  # No redacting of query strings
-            span.set_tag_str(http.URL, url)
-        else:  # Redact query strings
-            span.set_tag_str(http.URL, redact_url(url, config._obfuscation_query_string_pattern, query))
-    else:  # Not tagging query string in http.url
+    if not integration_config.http_tag_query_string:
+        span.set_tag_str(http.URL, strip_query_string(url))
+    elif config.global_query_string_obfuscation_disabled:
+        # TODO(munir): Remove this case when config.global_query_string_obfuscation_disabled is removed (v3.0),
+        # setting config._obfuscation_query_string_pattern to an empty string should be enough to disable obfuscation
```
Comment (mabdinur) [inline ddtrace/contrib/trace_utils.py:411]:
```suggestion
        # TODO(munir): This case exists for backwards compatibility. To remove query strings from URLs, 
        # users should set ``DD_TRACE_HTTP_CLIENT_TAG_QUERY_STRING=False``. This case should be 
        # removed when config.global_query_string_obfuscation_disabled is removed (v3.0).
```

File: ddtrace/settings/config.py [hunk 10]
```diff
@@ -567,18 +567,16 @@ def __init__(self):
...
-            except Exception:
-                log.warning("Invalid obfuscation pattern, disabling query string tracing", exc_info=True)
-                self.http_tag_query_string = False  # Disable query string tagging if malformed obfuscation pattern
+        try:
+            self._obfuscation_query_string_pattern = re.compile(
+                dd_trace_obfuscation_query_string_regexp.encode("ascii")
+            )
+        except Exception:
+            log.warning("Invalid obfuscation pattern, disabling query string tracing", exc_info=True)
+            self.http_tag_query_string = False  # Disable query string tagging if malformed obfuscation pattern
```
Comment (erikayasuda) [inline ddtrace/settings/config.py:579]: nit: Is there a reason we don't wrap this like:
```suggestion
      if not self.global_query_string_obfuscation_disabled:
        try:
            self._obfuscation_query_string_pattern = re.compile(
                dd_trace_obfuscation_query_string_regexp.encode("ascii")
            )
        except Exception:
            log.warning("Invalid obfuscation pattern, disabling query string tracing", exc_info=True)
            self.http_tag_query_string = False  # Disable query string tagging if malformed obfuscation pattern
```

**Review by erikayasuda — APPROVED** [review]

**Review by wconti27 — APPROVED** [review]

File: ddtrace/settings/config.py (hunk 10 above)
Comment (mabdinur) [reply ddtrace/settings/config.py:579]: The long term plan is to remove `config.global_query_string_obfuscation_disabled`. We want `self._obfuscation_query_string_pattern` to be consistent with with the environment variable (even when it's an empty string).

**Pull request merged**